src/julia-parser.scm

```diff
@@ -1254,22 +1262,50 @@
                      (body  (parse-block s)))
                 (expect-end s word)
                 (list word def body)))))
+
        ((abstract)
-        (list 'abstract (parse-subtype-spec s)))
-       ((type immutable)
-        (let ((immu? (eq? word 'immutable)))
-          (if (reserved-word? (peek-token s))
-              (error (string "invalid type name \"" (take-token s) "\"")))
-          (let ((sig (parse-subtype-spec s)))
-            (begin0 (list 'type (if (eq? word 'type) 'true 'false)
-                          sig (parse-block s))
-                    (expect-end s word)))))
+        (let ((ty (eq? (peek-token s) 'type)))
+          (if ty
+              (take-token s))
+          (let ((spec (parse-subtype-spec s)))
+            (if (not ty)
+                (syntax-deprecation s (string "abstract " (deparse spec))
+                                    (string "abstract type " (deparse spec) " end")))
+            (begin0 (list 'abstract spec)
+                    (if ty (expect-end s "abstract type"))))))
+       ((struct)
+        (begin (take-token s)
+               (parse-struct-def s #f word)))
+       ((mutable)
+        (if (not (eq? (peek-token s) 'struct))
+            (parse-call-chain s word #f)
+            (begin (take-token s)
+                   (parse-struct-def s #t word))))
+       ((primitive)
+        (if (not (eq? (peek-token s) 'type))
+            (parse-call-chain s word #f)
+            (begin (take-token s)
+                   (let* ((spec (with-space-sensitive (parse-subtype-spec s)))
+                          (nb   (with-space-sensitive (parse-cond s))))
+                     (begin0 (list 'bitstype nb spec)
+                             (expect-end s "primitive type"))))))
+       ;; deprecated type keywords
+       ((type)
+        (syntax-deprecation s "type" "mutable struct")
+        (parse-struct-def s #t word))
+       ((immutable)
+        (syntax-deprecation s "immutable" "struct")
+        (parse-struct-def s #f word))
```

Comment:
I don't think these should call `syntax-deprecations` in this release, because of the impossibility of `@compat` support.

For now, they should just be synonyms.  We can add the deprecation messages in the next release, at which point people can switch without breaking the previous release.